File: 501.find-mode-in-binary-search-tree.js
/*
 * @lc app=leetcode id=501 lang=javascript
 *
 * [501] Find Mode in Binary Search Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var findMode = function(root, data={}) {
    if (!root) return
    if (root.val in data) {
        data[root.val]++
    }
    else {
        data[root.val] = 1
    }
    findMode(root.left, data)
    findMode(root.right, data)
    return Object.keys(data).filter(key => data[key] === Math.max(...Object.values(data))).map(key => parseInt(key))
};
// @lc code=end

File: 501.find-mode-in-binary-search-tree.py
#
# @lc app=leetcode id=501 lang=python
#
# [501] Find Mode in Binary Search Tree
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def findMode(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """

        # Approach 1: Hash Table
        # Time  complexity: O(n)
        # Space complexity: O(n)
        def dfs(node):
            if node:
                count[node.val] += 1
                dfs(node.left)
                dfs(node.right)
        count = collections.Counter()
        dfs(root)
        max_count = max(count.values())
        return [k for k, v in count.items() if v == max_count]

       

# @lc code=end
